packages/blade/src/components/CounterInput/CounterInput.web.tsx

```diff
@@ -0,0 +1,229 @@
+import { assignWithoutSideEffects } from '~utils/assignWithoutSideEffects';
+import { makeAnalyticsAttribute } from '~utils/makeAnalyticsAttribute';
+import type { BladeElementRef } from '~utils/types';
+import { useControllableState } from '~utils/useControllable';
+import { getOuterMotionRef } from '~utils/getMotionRefs';
+import { StyledCounterInput } from './StyledCounterInput';
+import BaseBox from '~components/Box/BaseBox';
+import { FormLabel } from '~components/Form';
+import { useFormId } from '~components/Form/useFormId';
+import { useId } from '~utils/useId';
+import { useTheme } from '~components/BladeProvider';
+import { useBreakpoint } from '~utils';
+import { MinusIcon, PlusIcon } from '~components/Icons';
+import { COUNTER_INPUT_TOKEN } from './token';
+import { ProgressBar } from '~components/ProgressBar';
+import { CounterInputProvider } from './CounterInputContext';
+
+const _CounterInput = React.forwardRef<BladeElementRef, CounterInputProps>(
+  (
+    {
+      label,
+      accessibilityLabel,
+      labelPosition = 'top',
+      name,
+      value,
+      defaultValue,
+      min = 0,
+      max,
+      emphasis = 'subtle',
+      size = 'medium',
+      isLoading = false,
+      isDisabled = false,
+      onChange,
+      onFocus,
+      onBlur,
+      testID,
+      _motionMeta,
+      ...rest
+    },
+    ref,
+  ) => {
+    const [internalValue, setInternalValue] = useControllableState({
+      value: value,
+      defaultValue: defaultValue,
+      onChange: (newValue) => onChange?.({ value: newValue }),
+    });
+
+    const { inputId } = useFormId('counter-input');
+    const idBase = useId('counter-input');
+    const labelId = `${idBase}-label`;
+    const { theme } = useTheme();
+    const { matchedDeviceType } = useBreakpoint({ breakpoints: theme.breakpoints });
+    const isLabelLeftPositioned = labelPosition === 'left' && matchedDeviceType === 'desktop';
+    const emphasisTokens = COUNTER_INPUT_TOKEN.emphasis[emphasis];
+    const _isDisabled = isDisabled || isLoading;
+
+    const handleInputChange = useCallback(
+      ({ value: inputValue }: { value?: string }) => {
+        const numericValue = inputValue ? parseInt(inputValue, 10) : min;
+        const validValue = isNaN(numericValue) ? min : numericValue;
+
+        let constrainedValue = validValue;
+        if (constrainedValue < min) constrainedValue = min;
+        if (max !== undefined && constrainedValue > max) constrainedValue = max;
+        setInternalValue(() => constrainedValue);
+        onChange?.({ value: constrainedValue as number });
+      },
+      [min, max, onChange, name],
+    );
+
+    const handleIncrement = useCallback(() => {
+      if (_isDisabled) return;
+
+      const newValue = (internalValue ?? min) + 1;
+      const constrainedValue = max !== undefined ? Math.min(newValue, max) : newValue;
+
+      setInternalValue(() => constrainedValue);
+
+      onChange?.({ value: constrainedValue as number });
```

Comment:
onChange will already be called by the useControllableState? why are we calling it manually?